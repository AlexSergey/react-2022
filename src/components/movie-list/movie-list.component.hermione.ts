describe('Movie List Component', () => {
  it('primary', async function renderMovieList() {
    await this.browser.selectStory('components-movielistcomponent--movie-list');

    await this.browser.assertView('plain', 'body');
  });
});
