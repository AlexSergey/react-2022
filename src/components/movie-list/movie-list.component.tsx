import { useFetchMovies } from '../../hooks/fetch-movies/fetch-movies.hook';

import { MovieItemComponent } from './movie-item.component';

export const MovieListComponent = (): JSX.Element => {
  const { error, loading, data } = useFetchMovies();

  if (loading) {
    return <div>Loading</div>;
  }
  if (error) {
    return <div>Error</div>;
  }

  return (
    <div>
      {data?.map((movie) => (
        <MovieItemComponent key={movie.id} {...movie} />
      ))}
    </div>
  );
};
