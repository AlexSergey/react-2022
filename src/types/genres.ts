export enum Genres {
  Comedy = 'Comedy',
  Fantasy = 'Fantasy',
  Crime = 'Crime',
  Drama = 'Drama',
  Music = 'Music',
  Adventure = 'Adventure',
  History = 'History',
  Thriller = 'Thriller',
  Animation = 'Animation',
  Family = 'Family',
  Mystery = 'Mystery',
  Biography = 'Biography',
  Action = 'Action',
  'Film-Noir' = 'Film-Noir',
  Romance = 'Romance',
  'Sci-Fi' = 'Sci-Fi',
  War = 'War',
  Western = 'Western',
  Horror = 'Horror',
  Musical = 'Musical',
  Sport = 'Sport',
}
