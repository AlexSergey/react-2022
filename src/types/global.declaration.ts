declare module '*.bmp' {
  const src: string;
  export default src;
}

declare module '*.gif' {
  const src: string;
  export default src;
}

declare module '*.jpg' {
  const src: string;
  export default src;
}

declare module '*.jpeg' {
  const src: string;
  export default src;
}

declare module '*.png' {
  const src: string;
  export default src;
}

declare module '*.webp' {
  const src: string;
  export default src;
}

declare module '*.svg' {
  const src: string;
  export default src;
}

declare module '*.component.svg' {
  import * as React from 'react';

  const ReactComponent: React.FC<React.SVGProps<SVGSVGElement>>;
  export default ReactComponent;
}
