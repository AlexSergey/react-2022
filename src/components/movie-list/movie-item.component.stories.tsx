import { Meta, Story } from '@storybook/react';

import beetlejuice from '../../mocks/assets/beetlejuice.jpg';
import { Genres } from '../../types/genres';

import { IMovieItemComponent, MovieItemComponent } from './movie-item.component';

export default {
  argTypes: {
    actors: {
      description: 'List of actors separated by comma. Example: "Alec Baldwin, Geena Davis"',
    },
    genres: {
      description: `Enum of genres: ${(Object.keys(Genres) as (keyof typeof Genres)[])
        .map((genre) => `"${Genres[genre]}"`)
        .join(', ')}`,
    },
    posterUrl: {
      description: 'url or base64',
    },
  },
  component: MovieItemComponent,
  title: 'Components',
} as Meta<IMovieItemComponent>;

const Template: Story<IMovieItemComponent> = (props): JSX.Element => <MovieItemComponent {...props} />;

export const MovieItem: Story<IMovieItemComponent> = Template.bind({});

MovieItem.args = {
  actors: 'Alec Baldwin, Geena Davis, Annie McEnroe, Maurice Page',
  director: 'Tim Burton',
  genres: [Genres.Comedy, Genres.Fantasy],
  id: 1,
  plot: 'A couple of recently deceased ghosts contract the services of a "bio-exorcist" in order to remove the obnoxious new owners of their house.',
  posterUrl: beetlejuice,
  runtime: '92',
  title: 'Beetlejuice',
  year: '1988',
};
