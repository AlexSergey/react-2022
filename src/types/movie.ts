import { Genres } from './genres';

export interface IMovie {
  id: number;
  title: string;
  year: string;
  runtime: string;
  genres: Genres[];
  director: string;
  actors: string;
  plot: string;
  posterUrl: string;
}
