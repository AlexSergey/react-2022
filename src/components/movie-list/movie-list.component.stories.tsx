import { Story } from '@storybook/react';
import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';
import { useEffect, useRef } from 'react';

import cityOfGod from '../../mocks/assets/city-of-god.png';
import ratatouille from '../../mocks/assets/ratatouille.jpg';
import valkyrie from '../../mocks/assets/valkyrie.jpg';
import { Genres } from '../../types/genres';

import { MovieListComponent } from './movie-list.component';

export default {
  parameters: {
    docs: {
      description: {
        component:
          'Component for rendering movies. Presented in the form of cards with a film, a description and a banner.',
      },
    },
  },
  title: 'Components/MovieListComponent',
};

const useMockApi = (): void => {
  const apiMock = useRef<MockAdapter | false>(false);

  if (!apiMock.current) {
    apiMock.current = new MockAdapter(axios);

    apiMock.current.onGet('http://localhost:4444/api/movies').reply(200, [
      {
        actors: 'Tom Cruise, Kenneth Branagh, Bill Nighy, Tom Wilkinson',
        director: 'Bryan Singer',
        genres: [Genres.Drama, Genres.History, Genres.Thriller],
        id: 5,
        plot: 'A dramatization of the 20 July assassination and political coup plot by desperate renegade German Army officers against Hitler during World War II.',
        posterUrl: valkyrie,
        runtime: '121',
        title: 'Valkyrie',
        year: '2008',
      },
      {
        actors: 'Patton Oswalt, Ian Holm, Lou Romano, Brian Dennehy',
        director: 'Brad Bird, Jan Pinkava',
        genres: [Genres.Animation, Genres.Comedy, Genres.Family],
        id: 6,
        plot: 'A rat who can cook makes an unusual alliance with a young kitchen worker at a famous restaurant.',
        posterUrl: ratatouille,
        runtime: '111',
        title: 'Ratatouille',
        year: '2007',
      },
      {
        actors: 'Alexandre Rodrigues, Leandro Firmino, Phellipe Haagensen, Douglas Silva',
        director: 'Fernando Meirelles, KÃ¡tia Lund',
        genres: [Genres.Crime, Genres.Drama],
        id: 7,
        plot: 'Two boys growing up in a violent neighborhood of Rio de Janeiro take different paths: one becomes a photographer, the other a drug dealer.',
        posterUrl: cityOfGod,
        runtime: '130',
        title: 'City of God',
        year: '2002',
      },
    ]);
  }

  useEffect(() => {
    return (): void => {
      if (apiMock.current) {
        apiMock.current.reset();
      }
    };
  }, []);
};

const Template: Story<{}> = (): JSX.Element => {
  useMockApi();

  return <MovieListComponent />;
};

export const MovieList = Template.bind({});

MovieList.args = {};
