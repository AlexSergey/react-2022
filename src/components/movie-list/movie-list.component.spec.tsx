// eslint-disable-next-line import/no-extraneous-dependencies
import { expect } from '@jest/globals';
import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';
// eslint-disable-next-line import/no-extraneous-dependencies
import { act, create } from 'react-test-renderer';

import { sleep } from '../../helpers/sleep';
import { Genres } from '../../types/genres';

import { MovieListComponent } from './movie-list.component';

describe('Movies list interaction with api tests', () => {
  beforeAll(() => {
    const mock = new MockAdapter(axios);

    mock.onGet('http://localhost:4444/api/movies').reply(200, [
      {
        actors: 'Alec Baldwin, Geena Davis, Annie McEnroe, Maurice Page',
        director: 'Tim Burton',
        genres: [Genres.Comedy, Genres.Fantasy],
        id: 1,
        plot: 'A couple of recently deceased ghosts contract the services of a "bio-exorcist" in order to remove the obnoxious new owners of their house.',
        posterUrl:
          'https://images-na.ssl-images-amazon.com/images/M/MV5BMTUwODE3MDE0MV5BMl5BanBnXkFtZTgwNTk1MjI4MzE@._V1_SX300.jpg',
        runtime: '92',
        title: 'Beetlejuice',
        year: '1988',
      },
    ]);
  });

  it('MovieListComponent renders with one film', async () => {
    const movieList = create(<MovieListComponent />);

    await act(sleep(100));

    const tree = movieList.toJSON();

    expect(tree).toMatchSnapshot();
  });
});
