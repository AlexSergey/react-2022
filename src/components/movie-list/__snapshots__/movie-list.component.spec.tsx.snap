// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Movies list interaction with api tests ImageArea renders correctly 1`] = `
<div>
  <div
    className="max-w-4xl flex items-center h-auto lg:my-12 flex-wrap mx-auto my-24 lg:my-0"
  >
    <div
      className="w-full lg:w-3/5 rounded-lg lg:rounded-l-lg lg:rounded-r-none shadow-2xl bg-white dark:bg-slate-800 opacity-75 mx-6 lg:mx-0"
      id="profile"
    >
      <div
        className="p-4 md:p-12 text-center lg:text-left"
      >
        <div
          className="block lg:hidden rounded-full shadow-xl mx-auto -mt-16 h-48 w-48 bg-cover bg-center"
          style={
            {
              "backgroundImage": "url("https://images-na.ssl-images-amazon.com/images/M/MV5BMTUwODE3MDE0MV5BMl5BanBnXkFtZTgwNTk1MjI4MzE@._V1_SX300.jpg")",
            }
          }
        />
        <h1
          className="text-3xl font-bold pt-8 lg:pt-0 dark:text-white"
        >
          Beetlejuice
        </h1>
        <span
          className="text-1m font-bold pt-6 lg:pt-0"
        >
          1988
        </span>
        <div
          className="mx-auto lg:mx-0 w-4/5 pt-3 border-b-2 border-green-500 opacity-25"
        />
        <p
          className="pt-4 text-base font-bold flex items-center justify-center lg:justify-start dark:text-white"
        >
          <svg
            className="h-4 fill-current text-green-700 pr-4"
            viewBox="0 0 20 20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M9 12H1v6a2 2 0 0 0 2 2h14a2 2 0 0 0 2-2v-6h-8v2H9v-2zm0-1H0V5c0-1.1.9-2 2-2h4V2a2 2 0 0 1 2-2h4a2 2 0 0 1 2 2v1h4a2 2 0 0 1 2 2v6h-9V9H9v2zm3-8V2H8v1h4z"
            />
          </svg>
          Tim Burton
        </p>
        <p
          className="pt-2 text-gray-600 text-xs lg:text-sm flex items-center justify-center lg:justify-start dark:text-white"
        >
          <svg
            className="h-4 fill-current text-green-700 pr-4"
            viewBox="0 0 20 20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M10 20a10 10 0 1 1 0-20 10 10 0 0 1 0 20zm7.75-8a8.01 8.01 0 0 0 0-4h-3.82a28.81 28.81 0 0 1 0 4h3.82zm-.82 2h-3.22a14.44 14.44 0 0 1-.95 3.51A8.03 8.03 0 0 0 16.93 14zm-8.85-2h3.84a24.61 24.61 0 0 0 0-4H8.08a24.61 24.61 0 0 0 0 4zm.25 2c.41 2.4 1.13 4 1.67 4s1.26-1.6 1.67-4H8.33zm-6.08-2h3.82a28.81 28.81 0 0 1 0-4H2.25a8.01 8.01 0 0 0 0 4zm.82 2a8.03 8.03 0 0 0 4.17 3.51c-.42-.96-.74-2.16-.95-3.51H3.07zm13.86-8a8.03 8.03 0 0 0-4.17-3.51c.42.96.74 2.16.95 3.51h3.22zm-8.6 0h3.34c-.41-2.4-1.13-4-1.67-4S8.74 3.6 8.33 6zM3.07 6h3.22c.2-1.35.53-2.55.95-3.51A8.03 8.03 0 0 0 3.07 6z"
            />
          </svg>
          Alec Baldwin, Geena Davis, Annie McEnroe, Maurice Page
        </p>
        <p
          className="pt-8 text-sm dark:text-white"
        >
          A couple of recently deceased ghosts contract the services of a "bio-exorcist" in order to remove the obnoxious new owners of their house.
        </p>
      </div>
    </div>
    <div
      className="w-full lg:w-2/5"
    >
      <img
        className="rounded-none lg:rounded-lg shadow-2xl hidden lg:block"
        src="https://images-na.ssl-images-amazon.com/images/M/MV5BMTUwODE3MDE0MV5BMl5BanBnXkFtZTgwNTk1MjI4MzE@._V1_SX300.jpg"
      />
    </div>
  </div>
</div>
`;
